python_modules/dagster/dagster/_scheduler/scheduler.py

```diff
@@ -282,46 +286,72 @@ def launch_scheduled_runs(
         if code_location:
             for repo in code_location.get_repositories().values():
                 for schedule in repo.get_external_schedules():
                     selector_id = schedule.selector_id
+                    all_workspace_schedule_selector_ids.add(selector_id)
                     if schedule.get_current_instigator_state(
                         all_schedule_states.get(selector_id)
                     ).is_running:
-                        schedules[selector_id] = schedule
+                        running_schedules[selector_id] = schedule
+                    elif all_schedule_states.get(selector_id):
+                        schedule_state = all_schedule_states[selector_id]
+                        # If there is a DB row to update, see if we should still update the
+                        # last_iteration_timestamp
+                        _write_and_get_next_checkpoint_timestamp(
+                            instance,
+                            all_schedule_states[selector_id],
+                            cast(ScheduleInstigatorData, schedule_state.instigator_data),
+                            now_timestamp,
+                        )
+
         elif location_entry.load_error:
             error_locations.add(location_entry.origin.location_name)
 
-    # Remove any schedule states that were previously created with DECLARED_IN_CODE
-    # and can no longer be found in the workspace (so that if they are later added
-    # back again, their timestamps will start at the correct place)
+    # Remove any schedule states that were previously created and can no longer
+    # be found in the workspace (so that if they are later added back again,
+    # their timestamps will start at the correct place)
     states_to_delete = [
         schedule_state
         for selector_id, schedule_state in all_schedule_states.items()
-        if selector_id not in schedules
-        and schedule_state.status == InstigatorStatus.DECLARED_IN_CODE
+        if selector_id not in all_workspace_schedule_selector_ids
     ]
     for state in states_to_delete:
         location_name = state.origin.repository_origin.code_location_origin.location_name
-        # don't clean up auto running state if its location is an error state
-        if location_name not in error_locations:
-            logger.info(
-                f"Removing state for automatically running schedule {state.instigator_name} "
-                f"that is no longer present in {location_name}."
-            )
-            instance.delete_instigator_state(state.instigator_origin_id, state.selector_id)
 
-    if not schedules:
+        if location_name in error_locations:
+            # don't clean up state if its location is an error state
+            continue
+
+        if (
+            isinstance(state.instigator_data, ScheduleInstigatorData)
+            and end_datetime_utc.timestamp()
+            < (state.instigator_data.last_iteration_timestamp or 0)
```

Comment:
should this be a separate "and state.instigator_data.last_iteration_timestamp"? unless we go back in time to 1970, end_datetime_utc.timestamp() is never going to be < 0 + 12 hours right?